File: transfomer/canvas_border_autocrop/seeds.py
"""Multi-depth background seed sampling inside user ROI.

Samples along each side at several inset depths so workspace-background
frames (between outer UI and inner canvas) are not missed. Edge-only
sampling would preferentially hit UI chrome; deep-center sampling would
preferentially hit canvas.
"""

from __future__ import annotations

import numpy as np

from .config import DetectorConfig
from .features import CanvasFeatureMaps
from .types import IntRect, SeedPatch, SemanticSide


# Inset ratios along ROI short/long axes: shallow → mid → deeper frame bands
_BAND_RATIOS: tuple[float, ...] = (0.04, 0.09, 0.15, 0.22, 0.30)


def sample_background_seeds(
    features: CanvasFeatureMaps,
    user_roi: IntRect,
    cfg: DetectorConfig,
    dpi_scale: float,
) -> list[SeedPatch]:
    h, w = features.height, features.width
    short = min(user_roi.width, user_roi.height)
    size = cfg.seed_size_px(dpi_scale, short)
    half = size // 2
    seeds: list[SeedPatch] = []
    seed_id = 0
    n = cfg.seeds_per_side

    def _patch_stats(cx: int, cy: int) -> tuple[tuple[float, float, float], float, float] | None:
        x0 = max(0, cx - half)
        y0 = max(0, cy - half)
        x1 = min(w, x0 + size)
        y1 = min(h, y0 + size)
        if x1 - x0 < size // 2 or y1 - y0 < size // 2:
            return None
        lab = features.lab[y0:y1, x0:x1]
        g = features.gradient_magnitude[y0:y1, x0:x1]
        v = features.local_variance[y0:y1, x0:x1]
        thr = float(np.percentile(g, 70)) if g.size else 0.0
        grad_density = float(np.mean(g > thr)) if g.size else 1.0
        local_var = float(np.median(v)) if v.size else 1e9
        mean_lab = (
            float(np.median(lab[..., 0])),
            float(np.median(lab[..., 1])),
            float(np.median(lab[..., 2])),
        )
        return mean_lab, grad_density, local_var

    def _reject(grad_d: float, local_var: float, cx: int, cy: int) -> str:
        rcx = (user_roi.left + user_roi.right) * 0.5
        rcy = (user_roi.top + user_roi.bottom) * 0.5
        dx = abs(cx - rcx) / max(user_roi.width * 0.5, 1.0)
        dy = abs(cy - rcy) / max(user_roi.height * 0.5, 1.0)
        # Hard-exclude deep center (canvas body)
        if max(dx, dy) < cfg.seed_center_exclude_ratio * 0.55:
            return "too_central"
        if grad_d > cfg.seed_max_grad_density:
            return "high_gradient"
        if local_var > cfg.seed_max_local_var:
            return "high_variance"
        return ""

    points: list[tuple[SemanticSide, int, int]] = []

    for ratio in _BAND_RATIOS:
        inset_x = max(half + 1, int(round(user_roi.width * ratio)))
        inset_y = max(half + 1, int(round(user_roi.height * ratio)))
        if inset_x * 2 >= user_roi.width - size or inset_y * 2 >= user_roi.height - size:
            continue

        xs = np.linspace(
            user_roi.left + inset_x,
            user_roi.right - inset_x - 1,
            n,
            dtype=np.int32,
        )
        ys = np.linspace(
            user_roi.top + inset_y,
            user_roi.bottom - inset_y - 1,
            n,
            dtype=np.int32,
        )

        y_top = user_roi.top + inset_y
        y_bot = user_roi.bottom - inset_y - 1
        x_left = user_roi.left + inset_x
        x_right = user_roi.right - inset_x - 1

        for x in xs:
            points.append((SemanticSide.TOP, int(x), y_top))
            points.append((SemanticSide.BOTTOM, int(x), y_bot))
        for y in ys:
            points.append((SemanticSide.LEFT, x_left, int(y)))
            points.append((SemanticSide.RIGHT, x_right, int(y)))

        # Corner samples of this band
        points.append((SemanticSide.TOP, x_left, y_top))
        points.append((SemanticSide.TOP, x_right, y_top))
        points.append((SemanticSide.BOTTOM, x_left, y_bot))
        points.append((SemanticSide.BOTTOM, x_right, y_bot))

    # Deduplicate while preserving order
    seen: set[tuple[int, int]] = set()
    for side, cx, cy in points:
        key = (cx, cy)
        if key in seen:
            continue
        seen.add(key)
        if not (0 <= cx < w and 0 <= cy < h):
            continue
        if not user_roi.contains_point(cx, cy):
            continue
        stats = _patch_stats(cx, cy)
        if stats is None:
            seeds.append(SeedPatch(seed_id, side, cx, cy, size, (0.0, 0.0, 0.0), False, "oob"))
            seed_id += 1
            continue
        mean_lab, grad_d, local_var = stats
        reason = _reject(grad_d, local_var, cx, cy)
        seeds.append(
            SeedPatch(
                seed_id,
                side,
                cx,
                cy,
                size,
                mean_lab,
                accepted=reason == "",
                reject_reason=reason,
            )
        )
        seed_id += 1

    return seeds

File: screen_canvas_transform/native/src/seeds.cpp
#include "wb/seeds.hpp"

#include <algorithm>
#include <cmath>
#include <set>
#include <utility>

namespace wb {
namespace {

const float kBandRatios[] = {0.04f, 0.09f, 0.15f, 0.22f, 0.30f};

float MedianF(std::vector<float>& v) {
  if (v.empty()) return 0.f;
  const size_t m = v.size() / 2;
  std::nth_element(v.begin(), v.begin() + static_cast<std::ptrdiff_t>(m), v.end());
  return v[m];
}

}  // namespace

std::vector<SeedPatch> SampleBackgroundSeeds(const FeatureMaps& features, const IntRect& user_roi,
                                             const DetectorConfig& cfg, float dpi_scale) {
  const int h = features.height;
  const int w = features.width;
  const int short_side = std::min(user_roi.width(), user_roi.height());
  const int size = cfg.SeedSizePx(dpi_scale, short_side);
  const int half = size / 2;
  std::vector<SeedPatch> seeds;
  int seed_id = 0;
  const int n = cfg.seeds_per_side;

  auto patch_stats = [&](int cx, int cy, int patch_size, Lab& mean_lab, float& grad_density,
                         float& local_var) -> bool {
    const int x0 = std::max(user_roi.left, cx - patch_size/2);
    const int y0 = std::max(user_roi.top, cy - patch_size/2);
    const int x1 = std::min(user_roi.right, x0 + patch_size);
    const int y1 = std::min(user_roi.bottom, y0 + patch_size);
    if (x1 - x0 < patch_size || y1 - y0 < patch_size) return false;
    std::vector<float> L, A, B, gvals, vvals;
    for (int y = y0; y < y1; ++y) {
      for (int x = x0; x < x1; ++x) {
        const float* lab = features.lab.At(x, y);
        L.push_back(lab[0]);
        A.push_back(lab[1]);
        B.push_back(lab[2]);
        gvals.push_back(features.gradient_magnitude.At(x, y));
        vvals.push_back(features.local_variance.At(x, y));
      }
    }
    std::vector<float> gcopy = gvals;
    const float thr = Percentile(gcopy, 70.f);
    int above = 0;
    for (float g : gvals)
      if (g > thr) ++above;
    grad_density = gvals.empty() ? 1.f : static_cast<float>(above) / static_cast<float>(gvals.size());
    local_var = MedianF(vvals);
    mean_lab = {MedianF(L), MedianF(A), MedianF(B)};
    return true;
  };

  auto reject = [&](float grad_d, float local_var) -> std::string {
    // A coarse user selection may place real workspace background anywhere.
    // Geometry, not distance from the user ROI center, identifies its role.
    if (grad_d > cfg.seed_max_grad_density) return "high_gradient";
    if (local_var > cfg.seed_max_local_var) return "high_variance";
    return {};
  };

  std::vector<std::tuple<OuterSide, int, int>> points;
  for (float ratio : kBandRatios) {
    const int inset_x = std::max(half + 1, static_cast<int>(std::round(user_roi.width() * ratio)));
    const int inset_y = std::max(half + 1, static_cast<int>(std::round(user_roi.height() * ratio)));
    if (inset_x * 2 >= user_roi.width() - size || inset_y * 2 >= user_roi.height() - size) continue;

    const int y_top = user_roi.top + inset_y;
    const int y_bot = user_roi.bottom - inset_y - 1;
    const int x_left = user_roi.left + inset_x;
    const int x_right = user_roi.right - inset_x - 1;

    for (int i = 0; i < n; ++i) {
      const float t = (n == 1) ? 0.f : static_cast<float>(i) / static_cast<float>(n - 1);
      const int x = static_cast<int>(std::round(user_roi.left + inset_x +
                                                t * (user_roi.right - inset_x - 1 - (user_roi.left + inset_x))));
      const int y = static_cast<int>(std::round(user_roi.top + inset_y +
                                                t * (user_roi.bottom - inset_y - 1 - (user_roi.top + inset_y))));
      points.emplace_back(OuterSide::Top, x, y_top);
      points.emplace_back(OuterSide::Bottom, x, y_bot);
      points.emplace_back(OuterSide::Left, x_left, y);
      points.emplace_back(OuterSide::Right, x_right, y);
    }
    points.emplace_back(OuterSide::Top, x_left, y_top);
    points.emplace_back(OuterSide::Top, x_right, y_top);
    points.emplace_back(OuterSide::Bottom, x_left, y_bot);
    points.emplace_back(OuterSide::Bottom, x_right, y_bot);
  }

  // Small patches cover thin background strips that the proportional patches
  // straddle. Cover the whole user ROI, including its rim and center.
  const size_t large_patch_count = points.size();
  const int grid_step = std::clamp(short_side / 24, 3, 16);
  for (int y=user_roi.top+1; y<user_roi.bottom-1; y+=grid_step) {
    for (int x=user_roi.left+1; x<user_roi.right-1; x+=grid_step)
      points.emplace_back(OuterSide::Left,x,y);
    points.emplace_back(OuterSide::Right,user_roi.right-2,y);
  }
  for (int x=user_roi.left+1; x<user_roi.right-1; x+=grid_step)
    points.emplace_back(OuterSide::Bottom,x,user_roi.bottom-2);

  std::set<std::tuple<int, int, int>> seen;
  size_t point_index = 0;
  for (const auto& tup : points) {
    const int patch_size = point_index++ < large_patch_count ? size : 3;
    OuterSide side;
    int cx, cy;
    std::tie(side, cx, cy) = tup;
    if (!seen.insert({cx, cy, patch_size}).second) continue;
    if (cx < 0 || cx >= w || cy < 0 || cy >= h) continue;
    if (!user_roi.ContainsPoint(cx, cy)) continue;
    SeedPatch s;
    s.seed_id = seed_id++;
    s.side = side;
    s.x = cx;
    s.y = cy;
    s.size = patch_size;
    Lab mean{};
    float gd = 1.f, lv = 1e9f;
    if (!patch_stats(cx, cy, patch_size, mean, gd, lv)) {
      s.accepted = false;
      s.reject_reason = "oob";
      seeds.push_back(s);
      continue;
    }
    s.mean_lab = mean;
    s.reject_reason = reject(gd, lv);
    s.accepted = s.reject_reason.empty();
    seeds.push_back(s);
  }
  return seeds;
}

}  // namespace wb
